warehouse/api/billing.py

```diff
@@ -0,0 +1,140 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+
+import stripe
+
+from pyramid.httpexceptions import HTTPBadRequest, HTTPNoContent, HTTPNotFound
+from pyramid.view import view_config
+
+from warehouse.subscriptions.interfaces import IBillingService, ISubscriptionService
+from warehouse.subscriptions.models import StripeSubscriptionStatus
+
+
+def handle_billing_webhook_event(request, event):
+    billing_service = request.find_service(IBillingService, context=None)
+    subscription_service = request.find_service(ISubscriptionService, context=None)
+
+    match event["type"]:
+        # Occurs when a Checkout Session has been successfully completed.
+        case "checkout.session.completed":
+            checkout_session = event["data"]["object"]
+            # Get expanded checkout session object
+            checkout_session = billing_service.get_checkout_session(
+                checkout_session["id"],
+                # Provide mock_checkout_session used by MockStripeBillingService only.
+                mock_checkout_session=checkout_session,
+            )
+            status = checkout_session["status"]
+            customer_id = checkout_session["customer"]["id"]
+            subscription_id = checkout_session["subscription"]["id"]
+            if status != "complete":
+                raise HTTPBadRequest(f"Invalid checkout session status '{status}'")
+            if not customer_id:
+                raise HTTPBadRequest("Invalid customer ID")
+            if not subscription_id:
+                raise HTTPBadRequest("Invalid subscription ID")
+            if id := subscription_service.find_subscriptionid(subscription_id):
+                # Set subscription status to active.
+                subscription_service.update_subscription_status(
+                    id, StripeSubscriptionStatus.Active
+                )
+            else:
+                # Get expanded subscription object
+                subscription_items = checkout_session["subscription"]["items"]["data"]
+                # Activate subscription for customer.
+                for subscription_item in subscription_items:
+                    subscription_service.add_subscription(
+                        customer_id, subscription_id, subscription_item["id"]
+                    )
+                # Set subscription cycle to 1st of next month with no proration.
+                # https://stripe.com/docs/billing/subscriptions/billing-cycle#api-trials
+                stripe.Subscription.modify(
```

Comment:
Fixed the double invoicing in commit [3a1484c](https://github.com/pypi/warehouse/pull/11986/commits/3a1484c67ee798a19f550c322ecb6459bfa1a678) as discussed with @s-mm and @ewdurbin. Billing will now occur throughout the month depending on when subscriptions are created instead of having one big billing day.